### PR TITLE
imp: change ics20 events to emit token set

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #6347 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/transfer/types/token.go [hunk 1]
```diff
@@ -37,3 +37,24 @@ func (t Token) GetFullDenomPath() string {
 
 	return strings.Join(append(t.Trace, t.Denom), "/")
 }
+
+// Tokens is a set of Token
+type Tokens []Token
+
+// String prints out the tokens array as a string.
+// If the array is empty, an empty string is returned
+func (tokens Tokens) String() string {
```
Comment (colin-axner) [inline modules/apps/transfer/types/token.go:46]: shamelessly took from sdk.Coins.String()

**Comment by colin-axner** [comment]
No changelog needed, but migration entry should be added?

File: proto/ibc/applications/transfer/v2/packet.proto [hunk 2]
```diff
@@ -25,7 +27,7 @@ message FungibleTokenPacketData {
 // https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#data-structures
 message FungibleTokenPacketDataV2 {
   // the tokens to be transferred
-  repeated Token tokens = 1;
+  repeated Token tokens = 1 [(gogoproto.nullable) = false];
```
Comment (damiannolan) [inline proto/ibc/applications/transfer/v2/packet.proto:30]: 👍🏻

File: modules/apps/transfer/types/msgs.go [hunk 3]
```diff
@@ -108,7 +108,7 @@ func (msg MsgTransfer) ValidateBasic() error {
 }
 
 // GetTokens returns the tokens which will be transferred.
-func (msg MsgTransfer) GetTokens() []sdk.Coin {
+func (msg MsgTransfer) GetTokens() sdk.Coins {
```
Comment (damiannolan) [inline modules/apps/transfer/types/msgs.go:111]: 👍🏻

File: modules/apps/transfer/keeper/msg_server.go [hunk 4]
```diff
@@ -58,17 +58,14 @@ func (k Keeper) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.
...
-		))
-	}
+	k.Logger(ctx).Info("IBC fungible token transfer", "tokens", tokens, "sender", msg.Sender, "receiver", msg.Receiver)
+	events = append(events, sdk.NewEvent(
+		types.EventTypeTransfer,
+		sdk.NewAttribute(types.AttributeKeySender, msg.Sender),
+		sdk.NewAttribute(types.AttributeKeyReceiver, msg.Receiver),
+		sdk.NewAttribute(types.AttributeKeyTokens, tokens.String()),
+		sdk.NewAttribute(types.AttributeKeyMemo, msg.Memo),
+	))
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/msg_server.go:68]: should we still be creating a slice with underlying array `len(tokens)+1` and appending?  we're no longer ranging over `tokens`

I think the `[]sdk.Event` should actually be 2 now. One for the transfer event and one for the usual `EventTypeMessage` with module key/val.

**Review by damiannolan — CHANGES_REQUESTED** [review]
Awesome! Thanks for the improvements! I think we should make one change with slice/array allocation for events

File: modules/apps/transfer/keeper/msg_server.go (hunk 4 above)
Comment (colin-axner) [reply modules/apps/transfer/keeper/msg_server.go:68]: yes indeed. I'm just going to remove the `events` variable altogether and do direct emission

Comment (colin-axner) [reply modules/apps/transfer/keeper/msg_server.go]: done https://github.com/cosmos/ibc-go/pull/6348/commits/4ed0cb007616f73d893eb90ee001ccd7cb5d6f05

**Review by damiannolan — APPROVED** [review]
Beautiful! Nice one

**Pull request merged**